core/src/main/scala/org/apache/spark/sql/execution/ByteBufferHashMap.scala

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2017 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.nio.ByteBuffer
+
+import com.gemstone.gemfire.internal.shared.unsafe.UnsafeHolder
+
+import org.apache.spark.sql.execution.columnar.encoding.{ColumnAllocator, ColumnEncoding}
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.array.ByteArrayMethods
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * A HashMap implementation using a serialized ByteBuffer for key data and
+ * another one for value data. Key data is required to hold fixed-width
+ * values while the value data will be written back-to-back as new data
+ * is inserted into the map. Key data is stored in the following format:
+ * {{{
+ *     .------------------------ Offset into value data (4 bytes)
+ *     |   .-------------------- Hash code (4 bytes)
+ *     |   |
+ *     |   |        .----------- Fixed-width fields of key data
+ *     |   |        |        .-- Padding for 8 byte word alignment
+ *     V   V        V        V
+ *   +---+---+-------------+---+
+ *   |   |   | ... ... ... |   |
+ *   +---+---+-------------+---+
+ *    \-----/ \------------/\--/
+ *     header      body    padding
+ * }}}
+ * If key has variable length data, then it should be appended to the
+ * value data. The offset+hash code is read as a single long where LSB
+ * is used for hash code while MSB is used for offset, so the two can
+ * be reverse in actual memory layout on big-endian machines. Since there
+ * is no disk storage of this map so no attempt is made to have consistent
+ * endianness and thus the actual memory layout of the data.
+ */
+final class ByteBufferHashMap(initialCapacity: Int, val loadFactor: Double,
```

Comment:
Right now used only by dictionary encoder, but same will also be used instead of ObjectHashSet for off-heap support and reducing the objects (which should fix GC issues with TPCH Q21 when using hashjoin among others).

Will add more javadocs.